Added support for loop_old history variable.

aalok-thakkar opened the pull request:
1. Created a new predicate __CPROVER_loop_old which can be used in the loop invariant
   to keep track of the value of a variable at the start of the loop.
2. Modified the code contract to support this history variable.
3. Added regression tests.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

SaswatPadhi commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -246,6 +279,18 @@ void code_contractst::check_apply_loop_contracts(
 
+  // Reassign 'loop_old' history variables before the loop body
+  std::map<exprt, exprt> parameter2history_loop_old_2;
+  replace_history_parameter(
```
This doesn't look right.

1. It looks like you are replacing `__CPROVER_loop_old` within invariant_copy again, which would actually not replace anything, since those have been replaced before.

2. You should generate assignments here, not replacement. Basically, for each expression in `parameter2history_loop_old` (not `parameter2history_loop_old_2`, which shouldn't be needed) you should generate an assignment statement to update the stored value with the current value of the expression.

SaswatPadhi commented on regression/contracts/loop-history-variables-04/main.c:
```diff
@@ -0,0 +1,49 @@
+void main()
+{
+  int *x1, y1, z1;
+  *x1 = &z1;
```
```suggestion
  x1 = &z1;
```

SaswatPadhi reviewed (changes requested):
I think there is a major issue (commented below).

Adding more tests, where `__CPROVER_loop_old` would differ from `__CPROVER_loop_entry` would have helped catch the issue, so we should add those. For instance:

```c
  while(y1 > 0)
    __CPROVER_loop_invariant(
        *x1 + y1 == __CPROVER_loop_old(*x1) + __CPROVER_loop_old(y1)
    )
    {
      --y1;
      *x1 = *x1 + 1;
    }
```
Here `*x1` changes after each iteration, and so does `__CPROVER_loop_old(*x1)`, but `__CPROVER_loop_entry(*x1)` remains unchanged (= `z1`).

SaswatPadhi committed:
Added loop_old history variables for invariants

In this commit we introduce a new expression, `__CPROVER_loop_old`,
which can be used within loop invariants to track history of variables.
`__CPROVER_loop_old(x)` returns the initial value of a variable just
before the current iteration of a loop.

Co-authored-by: Aalok Thakkar <aalokt@dev-dsk-aalokt-1d-56ef39c6.us-east-1.amazon.com>
